test/Identity.IntegrationTest/Endpoints/IdentityServerTests.cs

```diff
@@ -0,0 +1,455 @@
+            var user = await database.Users
+                .FirstAsync(u => u.Email == username);
+
+            var context = await _factory.Server.PostAsync("/connect/token", new FormUrlEncodedContent(new Dictionary<string, string>
+            {
+                { "grant_type", "client_credentials" },
+                { "client_id", $"user.{user.Id}" },
+                { "client_secret", user.ApiKey },
+                { "scope", "api" },
+                { "DeviceIdentifier", deviceId },
+                { "DeviceType", DeviceTypeAsString(DeviceType.FirefoxBrowser) },
+                { "DeviceName", "firefox" },
+            }));
+
+            await AssertDefaultTokenBodyAsync(context, "api");
+        }
+
+        [Theory, BitAutoData]
+        public async Task TokenEndpoint_GrantTypeClientCredentials_AsOrganization_Success(Organization organization, OrganizationApiKey organizationApiKey)
+        {
+            var orgRepo = _factory.Services.GetRequiredService<IOrganizationRepository>();
+            organization = await orgRepo.CreateAsync(organization);
+            organizationApiKey.OrganizationId = organization.Id;
+            organizationApiKey.Type = OrganizationApiKeyType.Default;
+
+            var orgApiKeyRepo = _factory.Services.GetRequiredService<IOrganizationApiKeyRepository>();
+            await orgApiKeyRepo.CreateAsync(organizationApiKey);
+
+            var context = await _factory.Server.PostAsync("/connect/token", new FormUrlEncodedContent(new Dictionary<string, string>
+            {
+                { "grant_type", "client_credentials" },
+                { "client_id", $"organization.{organization.Id}" },
+                { "client_secret", organizationApiKey.ApiKey },
+                { "scope", "api.organization" },
+            }));
+
+            Assert.Equal(StatusCodes.Status200OK, context.Response.StatusCode);
+
+            await AssertDefaultTokenBodyAsync(context, "api.organization");
+        }
+
+        [Fact]
+        public async Task TokenEndpoint_GrantTypeClientCredentials_AsOrganization_BadOrgId_Fails()
+        {
+            var context = await _factory.Server.PostAsync("/connect/token", new FormUrlEncodedContent(new Dictionary<string, string>
+            {
+                { "grant_type", "client_credentials" },
+                { "client_id", "organization.bad_guid_zz&" },
+                { "client_secret", "something" },
+                { "scope", "api.organization" },
+            }));
+
+            Assert.Equal(StatusCodes.Status400BadRequest, context.Response.StatusCode);
+
+            var errorBody = await AssertHelper.AssertResponseTypeIsAsync<JsonDocument>(context);
+            var error = AssertHelper.AssertJsonProperty(errorBody.RootElement, "error", JsonValueKind.String).GetString();
+            Assert.Equal("invalid_client", error);
+        }
+
+        /// <summary>
+        /// This test currently does not test any code that is not covered by other tests but 
```

Comment:
But what if the `clientId` is `installation.`? Won't that return `idParts.Length = 0`?

Reply:
No the split will return back an array with 2 items `[ "installation", "" ]`